it/trace-context-leak/src/test/java/com/linecorp/armeria/common/TraceRequestContextLeakTest.java

```diff
@@ -0,0 +1,227 @@
+            final SafeCloseable leaked = ctx.push();
+            latch.countDown();
+            deferredClose.add(executor, leaked);
+        });
+
+        executor.next().execute(() -> {
+            //Leak happened on the first eventLoop doesn't make eventLoop when trying to push
+            final ServiceRequestContext anotherCtx = newCtx("/2");
+            try (SafeCloseable ignore = anotherCtx.push()) {
+                //ignore
+            } catch (Exception ex) {
+                //Not suppose to throw exception on the second event loop
+                isThrown.set(true);
+            } finally {
+                latch.countDown();
+            }
+        });
+
+        latch.await();
+        assertThat(isThrown).isFalse();
+
+        deferredClose.closeAll();
+    }
+
+    @Test
+    @SuppressWarnings("MustBeClosedChecker")
+    void multiThreadContextLeak() throws InterruptedException {
+        final AtomicBoolean isThrown = new AtomicBoolean(false);
+        final AtomicReference<Exception> exception = new AtomicReference<>();
+        final CountDownLatch latch = new CountDownLatch(3);
+        final DeferredClose deferredClose = new DeferredClose();
+
+        final EventLoopGroup executor =  eventLoopGroupExtension.get();
+
+        final ServiceRequestContext ctx = newCtx("/1");
+        final ServiceRequestContext anotherCtx = newCtx("/2");
+
+        final Executor ex1 = executor.next();
+        final Executor ex2 = executor.next();
+
+        ex1.execute(() -> {
+            final SafeCloseable leaked = ctx.push();
+            latch.countDown();
+            deferredClose.add(ex1, leaked);
+        });
+
+        ex2.execute(() -> {
+            try {
+                //intentional leak
+                final SafeCloseable leaked = anotherCtx.push();
+                deferredClose.add(ex2, leaked);
+            } catch (Exception ex) {
+                isThrown.set(true);
+            } finally {
+                latch.countDown();
+            }
+        });
+
+        assertThat(isThrown).isFalse();
+
+        ex1.execute(() -> {
+            try (SafeCloseable ignore = anotherCtx.push()) {
+                //ignore
+            } catch (Exception ex) {
+                isThrown.set(true);
+                exception.set(ex);
+            } finally {
+                latch.countDown();
+            }
+        });
+
+        latch.await();
+        assertThat(isThrown).isTrue();
+        assertThat(exception.get()).hasMessageContaining("RequestContext didn't popped");
+
+        deferredClose.closeAll();
+    }
+
+    private static ServiceRequestContext newCtx(String path) {
+        return ServiceRequestContext.builder(HttpRequest.of(HttpMethod.GET, path))
+                                    .build();
+    }
+
+    //Utility to clean up RequestContext leak after test
+    private static class DeferredClose {
+
+        private final AtomicReference<Map<Executor, SafeCloseable>> toClose;
```

Comment:
I am wondering why we need `AtomicReference` here? `ConcurrentHashMap` seems enough to me.

Reply:
You're right. I should've used `ConcurrentHashMap`😅.